rcl_yaml_param_parser/src/node_params.c

```diff
@@ -19,25 +19,87 @@
 rcutils_ret_t node_params_init(
   rcl_node_params_t * node_params,
   const rcutils_allocator_t allocator)
+{
+  return node_params_init_with_capacity(node_params, INIT_NUM_PARAMS_PER_NODE, allocator);
+}
+
+rcutils_ret_t node_params_init_with_capacity(
+  rcl_node_params_t * node_params,
+  size_t capacity,
+  const rcutils_allocator_t allocator)
 {
   RCUTILS_CHECK_ARGUMENT_FOR_NULL(node_params, RCUTILS_RET_INVALID_ARGUMENT);
   RCUTILS_CHECK_ALLOCATOR_WITH_MSG(
     &allocator, "invalid allocator", return RCUTILS_RET_INVALID_ARGUMENT);
+  if (capacity == 0) {
+    RCUTILS_SET_ERROR_MSG("capacity can't be zero");
+    return RCUTILS_RET_INVALID_ARGUMENT;
+  }
 
   node_params->parameter_names = allocator.zero_allocate(
-    MAX_NUM_PARAMS_PER_NODE, sizeof(char *), allocator.state);
+    capacity, sizeof(char *), allocator.state);
   if (NULL == node_params->parameter_names) {
+    RCUTILS_SET_ERROR_MSG("Failed to allocate memory for node parameter names");
     return RCUTILS_RET_BAD_ALLOC;
   }
 
   node_params->parameter_values = allocator.zero_allocate(
-    MAX_NUM_PARAMS_PER_NODE, sizeof(rcl_variant_t), allocator.state);
+    capacity, sizeof(rcl_variant_t), allocator.state);
   if (NULL == node_params->parameter_values) {
     allocator.deallocate(node_params->parameter_names, allocator.state);
     node_params->parameter_names = NULL;
+    RCUTILS_SET_ERROR_MSG("Failed to allocate memory for node parameter values");
+    return RCUTILS_RET_BAD_ALLOC;
+  }
+
+  node_params->num_params = 0U;
+  node_params->capacity_params = capacity;
+  return RCUTILS_RET_OK;
+}
+
+rcutils_ret_t node_params_reallocate(
+  rcl_node_params_t * node_params,
+  size_t new_capacity,
+  const rcutils_allocator_t allocator)
+{
+  RCUTILS_CHECK_ARGUMENT_FOR_NULL(node_params, RCUTILS_RET_INVALID_ARGUMENT);
+  RCUTILS_CHECK_ALLOCATOR_WITH_MSG(
+    &allocator, "invalid allocator", return RCUTILS_RET_INVALID_ARGUMENT);
+  // invalid if new_capacity is less than num_params
+  if (new_capacity < node_params->num_params) {
+    RCUTILS_SET_ERROR_MSG("invalid capacity");
```

Comment:
it would be nice to notify what size is expected for `new_capacity`.

Reply:
Thank you.